<h1>Data Cleaning / Munging</h1>

<p>In these notes we look at creating various plots based on a set of data.  We have a datafile called degreeClassification.csv (available on blackboard) which consists of the following columns:</p>
<ul>
<li>Academic Year:  The year that the degree was awarded.</li>
<li>Mode of study: Full-time, part-time or both.</li>
<li>Sex: Male, Female, Other or all.</li>
<li>Country of HE provider: England, Northern Ireland, Scotland, Wales or all.</li>
<li>Classification of First Degree: First, Upper Second, Lower Second or Third/Pass.</li>
<li>Number: The number of students achieving that degree.</li>
<li>Percentage:  The percentage of students achieving that degree.</li>
</ul>   

<p>We start the exercises by getting the data ready to be plotted</p>

<h2>Import PANDAs</h2>

In [3]:
import pandas as pd;

<h2>Read the data into a PANDAs DataFrame</h2>
<p>Read the 'degreeClassification.csv' in to a data frame called df, use the head method to output the first 20 rows of the datafile so that the structure of the dataframe can be explored.</p>

In [4]:
df = pd.read_csv('degreeClassification(1).csv')

<h2>Skipping Summary Rows</h2>
<p>Read the datafile and skip the rows that contain summary information.  Use the head method and output the first 20 rows.  The column names should be:
<ul>    
<li>Academic Year</li>
<li>Mode of study</li>
<li>Sex</li>
<li>Country of HE provider</li>
<li>Classification of First Degree</li>
<li>Number</li>
<li>Percentage</li>
</ul>    
</p>

In [19]:
df = pd.read_csv('degreeClassification(1).csv', skiprows=17)

column_names = [
    'Academic Year',
    'Mode of study',
    'Sex',
    'Country of HE provider',
    'Classification of First Degree',
    'Number',
    'Percentage',
]
df.columns = column_names

# display
df.head(20)


Academic Year Mode of study  Sex Country of HE provider  \
0        2013/14           All  All                    All   
1        2013/14           All  All                    All   
2        2013/14           All  All                    All   
3        2013/14           All  All                    All   
4        2013/14           All  All                    All   
5        2013/14           All  All                England   
6        2013/14           All  All                England   
7        2013/14           All  All                England   
8        2013/14           All  All                England   
9        2013/14           All  All                England   
10       2013/14           All  All       Northern Ireland   
11       2013/14           All  All       Northern Ireland   
12       2013/14           All  All       Northern Ireland   
13       2013/14           All  All       Northern Ireland   
14       2013/14           All  All       Northern Ireland   
15       2013/14           All  All               Scotland   
16       2013/14           All  All               Scotland   
17       2013/14           All  All               Scotland   
18       2013/14           All  All               Scotland   
19       2013/14           All  All               Scotland   

   Classification of First Degree  Number Percentage  
0                           First   79410        20%  
1                    Lower second   95965        24%  
2                      Third/Pass   21905         6%  
3                Total classified  395605        NaN  
4                    Upper second  198320        50%  
5                           First   68590        20%  
6                    Lower second   81665        24%  
7                      Third/Pass   19385         6%  
8                Total classified  339170        NaN  
9                    Upper second  169530        50%  
10                          First    1590        19%  
11                   Lower second    2215        26%  
12                     Third/Pass     285         3%  
13               Total classified    8560        NaN  
14                   Upper second    4470        52%  
15                          First    5125        21%  
16                   Lower second    5505        22%  
17                     Third/Pass     875         4%  
18               Total classified   24880        NaN  
19                   Upper second   13380        54%

<h2>Unique values in a Column</h2>
<p>Output the unique values for the column 'Mode of study' to give a general idea of the categorical data for this column</p>

In [20]:
df['Mode of study']

0             All
1             All
2             All
3             All
4             All
          ...    
1375    Part-time
1376    Part-time
1377    Part-time
1378    Part-time
1379    Part-time
Name: Mode of study, Length: 1380, dtype: object

<h2>Boolean Masks</h2>

<p>Create a boolean mask and then filter the data based on the categorical data value 'All' for 'Mode of Study'.  We do this in stages to help understand how the more complex aggregated instruction is built.</p>
<p>The overall goal is to apply a mask so that all the rows that contain values part-time or full-time are dropped from the dataframe.  We do this so that we don’t double count, i.e. A count in 'All' will also also be a count in either 'Part-time' or 'Full-time'.</p>

<ol>
<li>Create a mask to output True/False for each row, True if 'Mode of study' is equal to 'All' and False otherwise</li>

In [21]:
mask = (df['Mode of study'] ==  'All')

df[mask]

Academic Year Mode of study    Sex Country of HE provider  \
0          2013/14           All    All                    All   
1          2013/14           All    All                    All   
2          2013/14           All    All                    All   
3          2013/14           All    All                    All   
4          2013/14           All    All                    All   
...            ...           ...    ...                    ...   
1195       2017/18           All  Other                  Wales   
1196       2017/18           All  Other                  Wales   
1197       2017/18           All  Other                  Wales   
1198       2017/18           All  Other                  Wales   
1199       2017/18           All  Other                  Wales   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
3                  Total classified  395605        NaN  
4                      Upper second  198320        50%  
...                             ...     ...        ...  
1195                          First      10         ..  
1196                   Lower second       0         ..  
1197                     Third/Pass       0         ..  
1198               Total classified      20        NaN  
1199                   Upper second      10         ..  

[475 rows x 7 columns]

<ol start='2'>
<li>Assign the Boolean mask to a variable named mask</li>

In [22]:
mask = (df['Mode of study'] ==  'All')

<ol start='3'>
<li>Use the variable named mask to update the dataframe named df so that the dataframe is filtered based on the boolean mask</li>
<li>Output the unique values for the column 'Mode of study' to ensure that 'Full-time' and 'Part-time' have been filtered out (don't exist in the dataframe)</li>

In [23]:
df[mask]

Academic Year Mode of study    Sex Country of HE provider  \
0          2013/14           All    All                    All   
1          2013/14           All    All                    All   
2          2013/14           All    All                    All   
3          2013/14           All    All                    All   
4          2013/14           All    All                    All   
...            ...           ...    ...                    ...   
1195       2017/18           All  Other                  Wales   
1196       2017/18           All  Other                  Wales   
1197       2017/18           All  Other                  Wales   
1198       2017/18           All  Other                  Wales   
1199       2017/18           All  Other                  Wales   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
3                  Total classified  395605        NaN  
4                      Upper second  198320        50%  
...                             ...     ...        ...  
1195                          First      10         ..  
1196                   Lower second       0         ..  
1197                     Third/Pass       0         ..  
1198               Total classified      20        NaN  
1199                   Upper second      10         ..  

[475 rows x 7 columns]

<ol start='5'>
<li>Output the dataframe</li>

In [25]:
df = df[mask]
df

Academic Year Mode of study    Sex Country of HE provider  \
0          2013/14           All    All                    All   
1          2013/14           All    All                    All   
2          2013/14           All    All                    All   
3          2013/14           All    All                    All   
4          2013/14           All    All                    All   
...            ...           ...    ...                    ...   
1195       2017/18           All  Other                  Wales   
1196       2017/18           All  Other                  Wales   
1197       2017/18           All  Other                  Wales   
1198       2017/18           All  Other                  Wales   
1199       2017/18           All  Other                  Wales   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
3                  Total classified  395605        NaN  
4                      Upper second  198320        50%  
...                             ...     ...        ...  
1195                          First      10         ..  
1196                   Lower second       0         ..  
1197                     Third/Pass       0         ..  
1198               Total classified      20        NaN  
1199                   Upper second      10         ..  

[475 rows x 7 columns]

<h2>Boolean Masks Continued</h2>

<p>Create a boolean mask and then filter the data based on the categorical data 'All' for 'Sex'</p>
<p>Then create a boolean mask and then filter the data based on the categorical data 'All' for 'Country of HE provider'</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 1380 rows to 25 rows.</p>

In [77]:
mask = df['Sex'] == 'All'
df = df[mask]

mask = df['Country of HE provider'] == 'All'
df = df[mask]
df

KeyError: 'Sex'

<h2>More Boolean Masks</h2>

<p>We are not interested in rows that show the totals of classification for each year, filter these out.</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 25 rows to 20 rows.</p>

In [74]:
mask = df['Classification of First Degree'] != 'Total classified'
df[mask]


Academic Year Classification of First Degree  Number Percentage
1          2013/14                   Lower second   95965        24%
2          2013/14                     Third/Pass   21905         6%
4          2013/14                   Upper second  198320        50%
275        2014/15                          First   81640        22%
276        2014/15                   Lower second   85250        23%
277        2014/15                     Third/Pass   20340         5%
279        2014/15                   Upper second  183680        50%
550        2015/16                          First   88890        24%
551        2015/16                   Lower second   81595        22%
552        2015/16                     Third/Pass   19280         5%
554        2015/16                   Upper second  186565        50%
825        2016/17                          First  100945        26%
826        2016/17                   Lower second   79515        20%
827        2016/17                     Third/Pass   19025         5%
829        2016/17                   Upper second  192395        49%
1105       2017/18                          First  110475        28%
1106       2017/18                   Lower second   76210        19%
1107       2017/18                     Third/Pass   17830         4%
1109       2017/18                   Upper second  192425        48%

<h2>Dropping Columns</h2>
<p>Since columns 'Mode of study', 'Sex', and 'Country of HE provider' don\t provide us with any meaninful information (the all contain the data 'All'), drop these columns from the dataframe.</p>
<p>Output the dataframe to check the columns have been dropped successfully.</p>

In [78]:
df

Academic Year Classification of First Degree  Number Percentage
1          2013/14                   Lower second   95965        24%
2          2013/14                     Third/Pass   21905         6%
4          2013/14                   Upper second  198320        50%
275        2014/15                          First   81640        22%
276        2014/15                   Lower second   85250        23%
277        2014/15                     Third/Pass   20340         5%
279        2014/15                   Upper second  183680        50%
550        2015/16                          First   88890        24%
551        2015/16                   Lower second   81595        22%
552        2015/16                     Third/Pass   19280         5%
554        2015/16                   Upper second  186565        50%
825        2016/17                          First  100945        26%
826        2016/17                   Lower second   79515        20%
827        2016/17                     Third/Pass   19025         5%
829        2016/17                   Upper second  192395        49%
1105       2017/18                          First  110475        28%
1106       2017/18                   Lower second   76210        19%
1107       2017/18                     Third/Pass   17830         4%
1109       2017/18                   Upper second  192425        48%

<h2>Checking the Data Type of a Column</h2>
<p>Check the type of data that is stored for one of the values in the percentage column.</p>

In [101]:
df = pd.DataFrame(df)
type(df._get_value(4, 'Percentage'))

str

<h2>Cleaning and converting to an string data type to a number data type</h2>
<p>The percentage column is a string data type because of the % character.  Remove the % character and convert the values for this column to a numeric data type</p>

In [103]:
df

Percentage
1            24
2             6
4            50
275          22
276          23
277           5
279          50
550          24
551          22
552           5
554          50
825          26
826          20
827           5
829          49
1105         28
1106         19
1107          4
1109         48

<h2>New Dataframes</h2>
<p>Often it is useful to create seperate smaller dataframes, create four new dataframes named df_first, df_upp_sec, df_low_sec and df_third, to seperate the different degree classifications.</p>

Example:
df_first would output:
<pre>
      Academic Year   Classification of First Degree	Number	 Percentage
0	    2013/14	       First	                    79410	  20%
275	  2014/15	       First	                    81640	  22%
550	  2015/16	       First	                    88890	  24%
825	  2016/17	       First	                    100945	 26%
1105	 2017/18	       First	                    110475	 28%
</pre>

In [123]:
df = pd.read_csv('degreeClassification(1).csv', skiprows=17)
column_names = [
    'Academic Year',
    'Mode of study',
    'Sex',
    'Country of HE provider',
    'Classification of First Degree',
    'Number',
    'Percentage',
]
df.columns = column_names
mask_modeOfStudy = df['Mode of study'] == 'All'
mask_sex = df['Sex'] == 'All'
mask_countryOfHEProvider = df['Country of HE provider'] == 'All'
mask_classification = df['Classification of First Degree'] != 'Total classified'
df = df[mask_modeOfStudy]
df = df[mask_sex]
df = df[mask_countryOfHEProvider]
df = df[mask_classification]
# df = (df['Mode of study'] == 'All')
# df = df.drop(['Mode of study', 'Sex', 'Country of HE provider'], axis=1, inplace=False)
df = df.drop(['Mode of study', 'Sex', 'Country of HE provider'], axis=1, inplace= False)
mask_first = df['Classification of First Degree'] == 'First'
mask_upp_sec = df['Classification of First Degree'] == 'Upper second'
mask_low_sec = df['Classification of First Degree'] == 'Lower second'
mask_third = df['Classification of First Degree'] == 'Third/Pass'

df_first = df[mask_first]
df_upp_sec = df[mask_upp_sec]
df_low_sec = df[mask_low_sec]
df_third = df[mask_third]
df_first 

C:\Users\Maaz\AppData\Local\Temp\ipykernel_10896\3551920809.py:17: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df = df[mask_sex]
C:\Users\Maaz\AppData\Local\Temp\ipykernel_10896\3551920809.py:18: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df = df[mask_countryOfHEProvider]
C:\Users\Maaz\AppData\Local\Temp\ipykernel_10896\3551920809.py:19: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df = df[mask_classification]


Academic Year Classification of First Degree  Number Percentage
0          2013/14                          First   79410        20%
275        2014/15                          First   81640        22%
550        2015/16                          First   88890        24%
825        2016/17                          First  100945        26%
1105       2017/18                          First  110475        28%

<h2>Save the Dataframes to CSV files</h2>
<p>Save the dataframes df, df_first, df_upp_sec, df_low_sec and df_third to csv files with filesnames that match the dataframe name.

In [124]:
df_first.to_csv('df_first.csv')
df_upp_sec.to_csv('df_upp_sec.csv')
df_low_sec.to_csv('df_low_sec.csv')
df_third.to_csv('df_third.csv')